deep-learning/src/main/scala/com/microsoft/ml/spark/onnx/ONNXModel.scala

```diff
@@ -0,0 +1,496 @@
+
+  def setModelPayload(value: Array[Byte]): this.type = this.set(modelPayload, value)
+
+  def setModelLocation(path: String): this.type = {
+    val modelBytes = SparkContext.getOrCreate().binaryFiles(path).first()._2.toArray
+    this.setModelPayload(modelBytes)
+  }
+
+  val softMaxDict: MapParam[String, String] = new MapParam[String, String](
+    this,
+    "softMaxDict",
+    " A between output dataframe columns, the value column will be computed from taking the softmax of key column."
+  )
+
+  def setSoftMaxDict(value: Map[String, String]): this.type = set(softMaxDict, value)
+
+  def setSoftMaxDict(k: String, v: String): this.type = set(softMaxDict, Map(k -> v))
+
+  def getSoftMaxDict: Map[String, String] = get(softMaxDict).getOrElse(Map.empty)
+
+  val argMaxDict: MapParam[String, String] = new MapParam[String, String](
+    this,
+    "argMaxDict",
+    " A between output dataframe columns, the value column will be computed from taking the argmax of key column."
+  )
+
+  def setArgMaxDict(value: Map[String, String]): this.type = set(argMaxDict, value)
+
+  def setArgMaxDict(k: String, v: String): this.type = set(argMaxDict, Map(k -> v))
+
+  def getArgMaxDict: Map[String, String] = get(argMaxDict).getOrElse(Map.empty)
+
+  setDefault(
+    miniBatcher -> new FixedMiniBatchTransformer().setBatchSize(10) //scalastyle:ignore magic.number
+  )
+}
+
+object ONNXModel extends ComplexParamsReadable[ONNXModel] {
+  private[onnx] def initializeOrt(modelContent: Array[Byte]): OrtSession = {
+    val env: OrtEnvironment = OrtEnvironment.getEnvironment
+    val options = new SessionOptions()
+    options.setOptimizationLevel(OptLevel.BASIC_OPT)
+    env.createSession(modelContent, options)
+  }
+
+  private[onnx] def mapOnnxJavaTypeToDataType(javaType: OnnxJavaType): DataType = {
+    javaType match {
+      case OnnxJavaType.INT8 => ByteType
+      case OnnxJavaType.INT16 => ShortType
+      case OnnxJavaType.INT32 => IntegerType
+      case OnnxJavaType.INT64 => LongType
+      case OnnxJavaType.FLOAT => FloatType
+      case OnnxJavaType.DOUBLE => DoubleType
+      case OnnxJavaType.BOOL => BooleanType
+      case OnnxJavaType.STRING => StringType
+      case OnnxJavaType.UNKNOWN => BinaryType
+    }
+  }
+
+  private[onnx] def mapTensorInfoToDataType(tensorInfo: TensorInfo): DataType = {
+    val dataType = mapOnnxJavaTypeToDataType(tensorInfo.`type`)
+
+    def nestedArrayType(depth: Int, dataType: DataType): ArrayType = {
+      if (depth == 1)
+        ArrayType(dataType)
+      else
+        ArrayType(nestedArrayType(depth - 1, dataType))
+    }
+
+    if (tensorInfo.isScalar) {
+      dataType
+    } else if (tensorInfo.getShape.length == 1) {
+      // first dimension is assumed to be batch size.
+      dataType
+    } else {
+      nestedArrayType(tensorInfo.getShape.length - 1, dataType)
+    }
+  }
+
+  @tailrec
+  private[onnx] def mapValueInfoToDataType(valueInfo: ValueInfo): DataType = {
```

Comment:
Do you think that you could put a brief docstring on this. Just want to make sure we have the diff between Values, ONNXValues, ValueInfo, and Tensors explained somewhere as I cant quickly glean this from reading through the PR yet

Reply:
Done. I added some docstring to explain the ONNX object model (including ValueInfo, TensorInfo, SequenceInfo, etc.).